# __Pandas__
고수준의 자료 구조와 Python을 통한 데이터 처리 도구 <br>
Pandas의 기능 
* 축의 이름에 따라 정렬 가능한 자료 구조
* 색인 데이터 처리 기능 
* 시계열 데이터와 비시계열 데이터 처리 기능 
* 결측 데이터 처리 기능 
* SQL과 유사한 데이터 병합, 관계 연산 기능 
<br><br>

pandas를 활용하기 위해서는 가장 먼저 라이브러리 설치가 필요하다.

In [2]:
# install
!pip install pandas

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
# import
import pandas as pd

In [61]:
# Numpy도 불러옴
!pip install numpy
import numpy as np

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


## Series
: 1차원 배열에 해당하는 자료 구조로 index를 통한 접근 및 처리 가능 
* pd.Series(리스트)
* pd.Series(리스트, index=리스트)
* pd.Series(딕셔너리)

In [4]:
# 1. pd.Series(리스트) 이용
a1 = pd.Series([1,2,3,4])
print(a1)

0    1
1    2
2    3
3    4
dtype: int64


In [5]:
# 2. pd.Series(리스트, index=리스트) 이용
a2 = pd.Series([1,2,3,4], index=['a','b','c','d'])
print(a2)

a    1
b    2
c    3
d    4
dtype: int64


In [6]:
# 3. pd.Series(딕셔너리) 이용
dic = {'a': 100, 'b': 200, 'c': 300}
a3 = pd.Series(dic)
print(a3)

a    100
b    200
c    300
dtype: int64


In [7]:
# 4. 값, 인덱스 불러오기 
print(a3.values)  # 값 
print(a3.index) # 인덱스

[100 200 300]
Index(['a', 'b', 'c'], dtype='object')


In [8]:
# 5. 값이 존재하는지 확인
print( 'a' in a3 )
print( 100 in a3.values )

True
True


In [9]:
# 6. NaN 확인
pd.isnull(a3)

a    False
b    False
c    False
dtype: bool

## DataFrame
: 각 column에 따라 다른 자료 구조를 저장(행렬과 같은 구조) <br>
: 2차원 indexing으로 자료 접근, 처리 가능 
* pd.DataFrame(딕셔너리)
* pd.DataFrame(data, columns = object, index = object)

In [10]:
# 1. pd.DataFrame(딕셔너리) 이용 
dic2 = {
        'name': ['Kim', 'Lee', 'Kang', 'Kwon', 'Kim', 'Choi', 'Choi'],
        'age': [25, 28, 24, 26, 24, 23, 23],
        'sex': ['femle', 'male', 'female', 'female', 'female', 'female', 'female']
}  # 딕셔너리 만들기
df1 = pd.DataFrame(dic2)

df1

name  age     sex
0   Kim   25   femle
1   Lee   28    male
2  Kang   24  female
3  Kwon   26  female
4   Kim   24  female
5  Choi   23  female
6  Choi   23  female

In [11]:
# 2. pd.DataFrame(data, columns = object, index = object) 이용 
b = [[1,2,3], [4,5,6], [7,8,9]]  # 행렬 생성 
df2 = pd.DataFrame( b, columns= ['a', 'b', 'c'], index = ['d','e','f'])

df2

a  b  c
d  1  2  3
e  4  5  6
f  7  8  9

In [23]:
# 3. column에 해당 data가 없으면 NaN 처리
df3 = pd.DataFrame(dic2, columns= ['name', 'age', 'sex', 'state'])  # state 칼럼 새롭게 추가

df3

name  age     sex state
0   Kim   25   femle   NaN
1   Lee   28    male   NaN
2  Kang   24  female   NaN
3  Kwon   26  female   NaN
4   Kim   24  female   NaN
5  Choi   23  female   NaN
6  Choi   23  female   NaN

In [24]:
# 4. 동일한 크기의 배열 삽입 가능 
df3['state'] = ['Seoul','Seoul', 'Seoul', 'Seoul', 'Seoul', 'Suwon', 'Incheon']

df3

name  age     sex    state
0   Kim   25   femle    Seoul
1   Lee   28    male    Seoul
2  Kang   24  female    Seoul
3  Kwon   26  female    Seoul
4   Kim   24  female    Seoul
5  Choi   23  female    Suwon
6  Choi   23  female  Incheon

In [25]:
# 5. 열 조회: df.열이름 / 행 조회: df.loc[행이름]
print(df3.name)
print(df3.loc[2])
print(df3.loc[1][['name', 'age']])

0     Kim
1     Lee
2    Kang
3    Kwon
4     Kim
5    Choi
6    Choi
Name: name, dtype: object
name       Kang
age          24
sex      female
state     Seoul
Name: 2, dtype: object
name    Lee
age      28
Name: 1, dtype: object


## Pandas 핵심 기능
1. 재색인 기능 (reindexing)
  * 기존 데이터를 기반으로 새로운 index를 갖도록 데이터 생성
  * 기존 데이터의 index를 기반으로 재배치 및 병합
2. 특정 Rows or Columns 삭제
  * axis = 0: row, axis = 1: column
3. Indexing, Selection, Filtering
  * Numpy Array에서 사용한 Indexing과 동일
4. 두 객체 간의 산술 연산 
  * index가 동일한 경우에 연산이 이루어짐 
  * 누락된 index는 NaN 처리
5. Sorting and Ranking


In [26]:
# 1.  재색인 기능 예제
df2

a  b  c
d  1  2  3
e  4  5  6
f  7  8  9

In [27]:
# 인덱스 위치 변경
df4 = df2.reindex(['f', 'e', 'd'])
df4

a  b  c
f  7  8  9
e  4  5  6
d  1  2  3

In [28]:
# 새로운 인덱스 추가
df4.reindex(['d', 'e', 'f', 'g'], fill_value=0)

a  b  c
d  1  2  3
e  4  5  6
f  7  8  9
g  0  0  0

In [30]:
# 2. Rows or Columns 삭제
df3

name  age     sex    state
0   Kim   25   femle    Seoul
1   Lee   28    male    Seoul
2  Kang   24  female    Seoul
3  Kwon   26  female    Seoul
4   Kim   24  female    Seoul
5  Choi   23  female    Suwon
6  Choi   23  female  Incheon

In [32]:
# column 삭제
df3.drop('state', axis=1)

name  age     sex
0   Kim   25   femle
1   Lee   28    male
2  Kang   24  female
3  Kwon   26  female
4   Kim   24  female
5  Choi   23  female
6  Choi   23  female

In [33]:
# row 삭제
df3.drop(6, axis=0)

name  age     sex  state
0   Kim   25   femle  Seoul
1   Lee   28    male  Seoul
2  Kang   24  female  Seoul
3  Kwon   26  female  Seoul
4   Kim   24  female  Seoul
5  Choi   23  female  Suwon

In [36]:
# 여러 Column 삭제
df3.drop(['name', 'age'], axis=1)

sex    state
0   femle    Seoul
1    male    Seoul
2  female    Seoul
3  female    Seoul
4  female    Seoul
5  female    Suwon
6  female  Incheon

In [51]:
# 3. indexing, selection, filtering
# 기본적으로 행은 loc, 열은 그냥, 행 인덱스는 iloc
df3.loc[0]

name       Kim
age         25
sex      femle
state    Seoul
Name: 0, dtype: object

In [52]:
df3[0:3]

name  age     sex  state
0   Kim   25   femle  Seoul
1   Lee   28    male  Seoul
2  Kang   24  female  Seoul

In [38]:
df3['name']

0     Kim
1     Lee
2    Kang
3    Kwon
4     Kim
5    Choi
6    Choi
Name: name, dtype: object

In [39]:
df3[['name', 'age']]

name  age
0   Kim   25
1   Lee   28
2  Kang   24
3  Kwon   26
4   Kim   24
5  Choi   23
6  Choi   23

In [53]:
# 불린 값 이용
df3[df3['age'] >25]

name  age     sex  state
1   Lee   28    male  Seoul
3  Kwon   26  female  Seoul

In [45]:
# 행, 열 한번에 인덱싱
df3.loc[2, ['name', 'age']]

name    Kang
age       24
Name: 2, dtype: object

In [55]:
# 인덱싱 번호를 이용
df3.iloc[2]  # 그냥 df3[2]를 하면 에러

name       Kang
age          24
sex      female
state     Seoul
Name: 2, dtype: object

In [59]:
# 4. 두 객체 간의 산술 연산 
# 인덱스가 동일한 경우에 연산이 이루어짐
s1 = pd.Series([1,2,3,4,5], index = ['a', 'b', 'c', 'd','e'])
s2 = pd.Series([6,7,8,9], index=['a', 'c', 'e', 'f'])
s1 + s2 

a     7.0
b     NaN
c    10.0
d     NaN
e    13.0
f     NaN
dtype: float64

In [65]:
# 누락된 항목이 있는 object를 기준으로 method 적용
df1 = pd.DataFrame(np.arange(12).reshape((3,4)), columns=list('abcd'))
df2 = pd.DataFrame(np.arange(20).reshape((4,5)), columns=list('abcde'))
df1.add(df2, fill_value=0)  # NaN을 0으로 채운 뒤 덧셈 진행 (따라서 df1+df2와 결과값 다름)

a     b     c     d     e
0   0.0   2.0   4.0   6.0   4.0
1   9.0  11.0  13.0  15.0   9.0
2  18.0  20.0  22.0  24.0  14.0
3  15.0  16.0  17.0  18.0  19.0

In [66]:
# Series와 DataFrame 연산도 가능 
df2

a   b   c   d   e
0   0   1   2   3   4
1   5   6   7   8   9
2  10  11  12  13  14
3  15  16  17  18  19

In [67]:
s1

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [68]:
df2 + s1

a   b   c   d   e
0   1   3   5   7   9
1   6   8  10  12  14
2  11  13  15  17  19
3  16  18  20  22  24

In [76]:
# 5. Sorting and Ranking
s = pd.Series([1,3,5,7,9], index = ['e', 'd', 'c', 'b','a'])
s

e    1
d    3
c    5
b    7
a    9
dtype: int64

In [77]:
s.sort_index()

a    9
b    7
c    5
d    3
e    1
dtype: int64

In [78]:
s.rank()

e    1.0
d    2.0
c    3.0
b    4.0
a    5.0
dtype: float64

## 기본 전처리 과정

###파일 읽기 
아래의 예제에서는 .csv 파일을 사용하지만 excel, sql, json, parquet등의 파일도 읽을 수 있다.<br>
- pd.read_*('파일명.확장자') 

In [79]:
# csv 파일 읽기
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/BOAZ 20기 멘멘/titanic.csv') 
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [80]:
# 앞, 뒤의 n줄만 읽기
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [81]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

### DataFrame의 data type 확인

In [82]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

### DataFrame 정보 확인
* dataframe.info(): column개수, column이름, null값이 아닌 데이터의 개수, data type의 정보 포함

In [83]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**위의 결과로 알 수 있는 점**
- 전체 12개 열이 있다
- 전체 891개의 행이 있다
- age, cabin, embarked 열을 제외한 모든 열은 데이터가 다 있다.
- 83.7 kb 정도 메모리를 차지한다
- ...

### 여러 기능을 활용해 데이터 살펴보기

In [84]:
# 1. 특정 열의 shape 확인
df['Age'].shape

(891,)

In [85]:
# 2. 여러 개 열 선택 
df[['Age', 'Sex']]

Age     Sex
0    22.0    male
1    38.0  female
2    26.0  female
3    35.0  female
4    35.0    male
..    ...     ...
886  27.0    male
887  19.0  female
888   NaN  female
889  26.0    male
890  32.0    male

[891 rows x 2 columns]

In [86]:
# 3. 비교 연산자 이용 
df[df['Age'] > 35] 

PassengerId  Survived  Pclass  \
1              2         1       1   
6              7         0       1   
11            12         1       1   
13            14         0       3   
15            16         1       2   
..           ...       ...     ...   
865          866         1       2   
871          872         1       1   
873          874         0       3   
879          880         1       1   
885          886         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
..                                                 ...     ...   ...    ...   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
6        0     17463  51.8625   E46        S  
11       0    113783  26.5500  C103        S  
13       5    347082  31.2750   NaN        S  
15       0    248706  16.0000   NaN        S  
..     ...       ...      ...   ...      ...  
865      0    236852  13.0000   NaN        S  
871      1     11751  52.5542   D35        S  
873      0    345765   9.0000   NaN        S  
879      1     11767  83.1583   C50        C  
885      5    382652  29.1250   NaN        Q  

[217 rows x 12 columns]

In [89]:
# 4. 특정 값에 해당하는 경우만 추출 
df[df['Pclass'].isin([2, 3])]

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[675 rows x 12 columns]

In [90]:
# 5. 결측치가 아닌 값만 포함
df[df['Age'].notna()]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

In [93]:
# 6. 결측치만 포함
df[df['Age'].isnull()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

In [94]:
# 7. 결측치가 있는 행이 몇 개인지 확인
df['Age'].isnull().sum()

177

In [95]:
# 8. 특정 row, column의 데이터 바꾸기
df.iloc[0:3, 3] = 'anonymous'
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                                     anonymous    male  22.0      1      0   
1                                     anonymous  female  38.0      1      0   
2                                     anonymous  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

In [96]:
# 9. 계산된 값을 열에 추가 
df['Age2'] = df['Age'] * 2
df.head() 

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                                     anonymous    male  22.0      1      0   
1                                     anonymous  female  38.0      1      0   
2                                     anonymous  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  Age2  
0         A/5 21171   7.2500   NaN        S  44.0  
1          PC 17599  71.2833   C85        C  76.0  
2  STON/O2. 3101282   7.9250   NaN        S  52.0  
3            113803  53.1000  C123        S  70.0  
4            373450   8.0500   NaN        S  70.0

In [97]:
# 10. column 이름 변경 
df_rename = df.rename(columns = {'Name': 'name', 'Age': 'age'}) 
df_rename.head() 

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           name     Sex   age  SibSp  Parch  \
0                                     anonymous    male  22.0      1      0   
1                                     anonymous  female  38.0      1      0   
2                                     anonymous  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  Age2  
0         A/5 21171   7.2500   NaN        S  44.0  
1          PC 17599  71.2833   C85        C  76.0  
2  STON/O2. 3101282   7.9250   NaN        S  52.0  
3            113803  53.1000  C123        S  70.0  
4            373450   8.0500   NaN        S  70.0

In [98]:
# 11. 통계량 계산 
df[['Age', 'Fare']].median() 

Age     28.0000
Fare    14.4542
dtype: float64

In [99]:
df[['Age', 'Fare']].describe() 

Age        Fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

In [100]:
# 12. 카테고리별 열들의 통계량
df.groupby('Sex')['Age'].mean() 

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [105]:
df.groupby(['Pclass', 'Sex'])['Age'].sum()

Pclass  Sex   
1       female    2942.00
        male      4169.42
2       female    2125.50
        male      3043.33
3       female    2218.50
        male      6706.42
Name: Age, dtype: float64

In [101]:
# 13. 특정 열에서 카테고리별로 카운트 
df['Sex'].value_counts()  # == df.groupby('Sex')['Sex'].count()

male      577
female    314
Name: Sex, dtype: int64

In [102]:
# 14. 특정열의 고유값 구하기
df['Sex'].unique() 

array(['male', 'female'], dtype=object)

In [103]:
# 15. 나이 오름차순으로 정렬 
df.sort_values(by = 'Age').head(10)

PassengerId  Survived  Pclass                             Name     Sex  \
803          804         1       3  Thomas, Master. Assad Alexander    male   
755          756         1       2        Hamalainen, Master. Viljo    male   
644          645         1       3           Baclini, Miss. Eugenie  female   
469          470         1       3    Baclini, Miss. Helene Barbara  female   
78            79         1       2    Caldwell, Master. Alden Gates    male   
831          832         1       2  Richards, Master. George Sibley    male   
305          306         1       1   Allison, Master. Hudson Trevor    male   
827          828         1       2            Mallet, Master. Andre    male   
381          382         1       3      Nakid, Miss. Maria ("Mary")  female   
164          165         0       3     Panula, Master. Eino Viljami    male   

      Age  SibSp  Parch           Ticket      Fare    Cabin Embarked  Age2  
803  0.42      0      1             2625    8.5167      NaN        C  0.84  
755  0.67      1      1           250649   14.5000      NaN        S  1.34  
644  0.75      2      1             2666   19.2583      NaN        C  1.50  
469  0.75      2      1             2666   19.2583      NaN        C  1.50  
78   0.83      0      2           248738   29.0000      NaN        S  1.66  
831  0.83      1      1            29106   18.7500      NaN        S  1.66  
305  0.92      1      2           113781  151.5500  C22 C26        S  1.84  
827  1.00      0      2  S.C./PARIS 2079   37.0042      NaN        C  2.00  
381  1.00      0      2             2653   15.7417      NaN        C  2.00  
164  1.00      4      1          3101295   39.6875      NaN        S  2.00

In [104]:
# 16. plcass가 3인 row만 필터링
df[df['Pclass'] == 3] 

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                                 anonymous   
2              3         1       3                                 anonymous   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S   
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
4      male  35.0      0      0            373450   8.0500   NaN        S   
5      male   NaN      0      0            330877   8.4583   NaN        Q   
7      male   2.0      3      1            349909  21.0750   NaN        S   
..      ...   ...    ...    ...               ...      ...   ...      ...   
882  female  22.0      0      0              7552  10.5167   NaN        S   
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S   
885  female  39.0      0      5            382652  29.1250   NaN        Q   
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S   
890    male  32.0      0      0            370376   7.7500   NaN        Q   

     Age2  
0    44.0  
2    52.0  
4    70.0  
5     NaN  
7     4.0  
..    ...  
882  44.0  
884  50.0  
885  78.0  
888   NaN  
890  64.0  

[491 rows x 13 columns]

In [106]:
 # 17. 두 데이터 합치기 
df1 = df.iloc[:400]
df2 = df.iloc[400:] 

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                       anonymous    male  22.0      1      0   
1                                       anonymous  female  38.0      1      0   
2                                       anonymous  female  26.0      0      0   
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                        Allen, Mr. William Henry    male  35.0      0      0   
..                                            ...     ...   ...    ...    ...   
886                         Montvila, Rev. Juozas    male  27.0      0      0   
887                  Graham, Miss. Margaret Edith  female  19.0      0      0   
888      Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1      2   
889                         Behr, Mr. Karl Howell    male  26.0      0      0   
890                           Dooley, Mr. Patrick    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  Age2  
0           A/5 21171   7.2500   NaN        S  44.0  
1            PC 17599  71.2833   C85        C  76.0  
2    STON/O2. 3101282   7.9250   NaN        S  52.0  
3              113803  53.1000  C123        S  70.0  
4              373450   8.0500   NaN        S  70.0  
..                ...      ...   ...      ...   ...  
886            211536  13.0000   NaN        S  54.0  
887            112053  30.0000   B42        S  38.0  
888        W./C. 6607  23.4500   NaN        S   NaN  
889            111369  30.0000  C148        C  52.0  
890            370376   7.7500   NaN        Q  64.0  

[891 rows x 13 columns]

In [108]:
# axis = 0: 행방향
df_new = pd.concat([df1, df2], axis = 0) 
df_new

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                       anonymous    male  22.0      1      0   
1                                       anonymous  female  38.0      1      0   
2                                       anonymous  female  26.0      0      0   
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                        Allen, Mr. William Henry    male  35.0      0      0   
..                                            ...     ...   ...    ...    ...   
886                         Montvila, Rev. Juozas    male  27.0      0      0   
887                  Graham, Miss. Margaret Edith  female  19.0      0      0   
888      Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1      2   
889                         Behr, Mr. Karl Howell    male  26.0      0      0   
890                           Dooley, Mr. Patrick    male  32.0      0      0   

               Ticket     Fare Cabin Embarked  Age2  
0           A/5 21171   7.2500   NaN        S  44.0  
1            PC 17599  71.2833   C85        C  76.0  
2    STON/O2. 3101282   7.9250   NaN        S  52.0  
3              113803  53.1000  C123        S  70.0  
4              373450   8.0500   NaN        S  70.0  
..                ...      ...   ...      ...   ...  
886            211536  13.0000   NaN        S  54.0  
887            112053  30.0000   B42        S  38.0  
888        W./C. 6607  23.4500   NaN        S   NaN  
889            111369  30.0000  C148        C  52.0  
890            370376   7.7500   NaN        Q  64.0  

[891 rows x 13 columns]

In [107]:
# axis = 1: 열방향
df_new2 = pd.concat([df1, df2], axis = 1)
df_new2

PassengerId  Survived  Pclass  \
0            1.0       0.0     3.0   
1            2.0       1.0     1.0   
2            3.0       1.0     3.0   
3            4.0       1.0     1.0   
4            5.0       0.0     3.0   
..           ...       ...     ...   
886          NaN       NaN     NaN   
887          NaN       NaN     NaN   
888          NaN       NaN     NaN   
889          NaN       NaN     NaN   
890          NaN       NaN     NaN   

                                             Name     Sex   Age  SibSp  Parch  \
0                                       anonymous    male  22.0    1.0    0.0   
1                                       anonymous  female  38.0    1.0    0.0   
2                                       anonymous  female  26.0    0.0    0.0   
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0    0.0   
4                        Allen, Mr. William Henry    male  35.0    0.0    0.0   
..                                            ...     ...   ...    ...    ...   
886                                           NaN     NaN   NaN    NaN    NaN   
887                                           NaN     NaN   NaN    NaN    NaN   
888                                           NaN     NaN   NaN    NaN    NaN   
889                                           NaN     NaN   NaN    NaN    NaN   
890                                           NaN     NaN   NaN    NaN    NaN   

               Ticket     Fare  ...                                      Name  \
0           A/5 21171   7.2500  ...                                       NaN   
1            PC 17599  71.2833  ...                                       NaN   
2    STON/O2. 3101282   7.9250  ...                                       NaN   
3              113803  53.1000  ...                                       NaN   
4              373450   8.0500  ...                                       NaN   
..                ...      ...  ...                                       ...   
886               NaN      NaN  ...                     Montvila, Rev. Juozas   
887               NaN      NaN  ...              Graham, Miss. Margaret Edith   
888               NaN      NaN  ...  Johnston, Miss. Catherine Helen "Carrie"   
889               NaN      NaN  ...                     Behr, Mr. Karl Howell   
890               NaN      NaN  ...                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin  Embarked  Age2  
0       NaN   NaN    NaN    NaN         NaN    NaN   NaN       NaN   NaN  
1       NaN   NaN    NaN    NaN         NaN    NaN   NaN       NaN   NaN  
2       NaN   NaN    NaN    NaN         NaN    NaN   NaN       NaN   NaN  
3       NaN   NaN    NaN    NaN         NaN    NaN   NaN       NaN   NaN  
4       NaN   NaN    NaN    NaN         NaN    NaN   NaN       NaN   NaN  
..      ...   ...    ...    ...         ...    ...   ...       ...   ...  
886    male  27.0    0.0    0.0      211536  13.00   NaN         S  54.0  
887  female  19.0    0.0    0.0      112053  30.00   B42         S  38.0  
888  female   NaN    1.0    2.0  W./C. 6607  23.45   NaN         S   NaN  
889    male  26.0    0.0    0.0      111369  30.00  C148         C  52.0  
890    male  32.0    0.0    0.0      370376   7.75   NaN         Q  64.0  

[891 rows x 26 columns]

In [ ]:
# 데이터를 합칠 때, join과 merge함수를 이용할 수도 있음 

### 데이터 저장하기



In [111]:
# new_df.xlsx 파일로 저장
# 시트의 네임은 passengers
# 인덱스는 포함하지 않고 내보내기
df.to_excel('new_df.xlsx', sheet_name = 'passengers', index = False) 

In [112]:
# 다시 불러와서 저장됐는 지 확인
new_df = pd.read_excel('new_df.xlsx', sheet_name = 'passengers')
new_df.head() 

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                                     anonymous    male  22.0      1      0   
1                                     anonymous  female  38.0      1      0   
2                                     anonymous  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  Age2  
0         A/5 21171   7.2500   NaN        S  44.0  
1          PC 17599  71.2833   C85        C  76.0  
2  STON/O2. 3101282   7.9250   NaN        S  52.0  
3            113803  53.1000  C123        S  70.0  
4            373450   8.0500   NaN        S  70.0

## 과제 
원하는 데이터셋을 불러와 데이터 형태 살펴보기! (10개 이상의 서로 다른 기능 사용)